mlrun/datastore/azure_blob.py

```diff
@@ -53,36 +56,76 @@ def get_storage_options(self):
 
     def upload(self, key, src_path):
-        # Need to strip leading / from key
-        blob_client = self.bsc.get_blob_client(container=self.endpoint, blob=key[1:])
-        with open(src_path, "rb") as data:
-            blob_client.upload_blob(data, overwrite=True)
+        if self.bsc:
+            # Need to strip leading / from key
+            blob_client = self.bsc.get_blob_client(
+                container=self.endpoint, blob=key[1:]
+            )
+            with open(src_path, "rb") as data:
+                blob_client.upload_blob(data, overwrite=True)
+        else:
+            remote_path = f"{self.endpoint}{key}"
+            self._filesystem.put_file(src_path, remote_path)
 
     def get(self, key, size=None, offset=0):
-        blob_client = self.bsc.get_blob_client(container=self.endpoint, blob=key[1:])
-        size = size if size else None
-        return blob_client.download_blob(offset, size).readall()
+        if self.bsc:
+            blob_client = self.bsc.get_blob_client(
+                container=self.endpoint, blob=key[1:]
+            )
+            size = size if size else None
+            return blob_client.download_blob(offset, size).readall()
+        else:
+            blob = self._filesystem.cat_file(key, start=offset)
+            return blob
 
     def put(self, key, data, append=False):
-        blob_client = self.bsc.get_blob_client(container=self.endpoint, blob=key[1:])
-        # Note that append=True is not supported. If the blob already exists, this call will fail
-        blob_client.upload_blob(data, overwrite=True)
+        if self.bsc:
+            blob_client = self.bsc.get_blob_client(
+                container=self.endpoint, blob=key[1:]
+            )
+            # Note that append=True is not supported. If the blob already exists, this call will fail
+            blob_client.upload_blob(data, overwrite=True)
+        else:
+            path = f"{self.endpoint}{key}"
+            with self._filesystem.open(path, "wb") as f:
+                f.write(data)
 
     def stat(self, key):
-        blob_client = self.bsc.get_blob_client(container=self.endpoint, blob=key[1:])
-        props = blob_client.get_blob_properties()
-        size = props.size
-        modified = props.last_modified
+        if self.bsc:
+            blob_client = self.bsc.get_blob_client(
+                container=self.endpoint, blob=key[1:]
+            )
+            props = blob_client.get_blob_properties()
+            size = props.size
+            modified = props.last_modified
+        else:
+            path = f"{self.endpoint}{key}"
+            files = self._filesystem.ls(path, detail=True)
+            if len(files) == 0 and files[0]["kind"] == "file":
+                size = files[0]["size"]
+                modified = files[0]["last_modified"]
+            elif len(files) == 0 and files[0]["kind"] == "directory":
+                raise FileNotFoundError("Operation expects a file not a direcdtory!")
```

Comment:
```suggestion
            elif len(files) == 1 and files[0]["kind"] == "directory":
                raise FileNotFoundError("Operation expects a file not a directory!")
```

Are we guaranteed that when doing an `ls` operation on a directory, the first item returned will be the directory itself? I would actually expect an `ls` operation to not return the directory at all (in Linux it may return `.` in some cases, but I don't know how this works for `adlfs`).

I think it might be enough in this case to simply fail in this case:
```suggestion
            else:
                raise FileNotFoundError("Operation expects a single file, not a directory!")
```

Since the only valid case is covered before.